File: packages/python/src/softschema/core/envelope.py
"""Portable artifact-envelope selection."""

from __future__ import annotations

from typing import Any


class EnvelopeAmbiguityError(ValueError):
    """Multiple top-level payload candidates; the envelope must be designated."""

    def __init__(self, candidates: list[str]) -> None:
        self.candidates = candidates
        super().__init__(
            "multiple top-level frontmatter keys; designate the softschema payload "
            f"(candidates: {', '.join(candidates)})"
        )


def infer_envelope_key(frontmatter: dict[str, Any]) -> str | None:
    """Return the only non-metadata key, or require explicit disambiguation."""
    candidates = [key for key in frontmatter if key != "softschema"]
    if not candidates:
        return None
    if len(candidates) == 1:
        return candidates[0]
    raise EnvelopeAmbiguityError(candidates)

File: packages/typescript/src/core/envelope.ts
/** Runtime-neutral artifact-envelope selection. */

export class EnvelopeAmbiguityError extends Error {
  readonly candidates: string[];

  constructor(candidates: string[]) {
    super(
      "multiple top-level frontmatter keys; designate the softschema payload " +
        `(candidates: ${candidates.join(", ")})`,
    );
    this.candidates = [...candidates];
  }
}

/** Return the only non-metadata key, or require explicit disambiguation. */
export function inferEnvelopeKey(frontmatter: Record<string, unknown>): string | null {
  const candidates = Object.keys(frontmatter).filter((key) => key !== "softschema");
  if (candidates.length === 0) return null;
  if (candidates.length === 1) return candidates[0] ?? null;
  throw new EnvelopeAmbiguityError(candidates);
}
